blog/2021-03-16-BatchSolvingWithAmq.adoc

```diff
@@ -0,0 +1,182 @@
+Let’s just very briefly revisit the problem domain and its constraints.
+
+In the school timetabling problem, the goal is to assign each lesson to a room and a timeslot.
+To use the OptaPlanner vocabulary, the `Lesson` is a planning entity and its references to the `Room` and the `Timeslot` are planning variables.
+
+image::schoolTimetablingClassDiagram.png[School Timetabling class diagram]
+
+The `TimeTableConstraintProvider` defines the following constraints on how the lessons should be assigned to timeslots and rooms:
+
+* A room can have at most one lesson at the same time (hard).
+* A teacher can teach at most one lesson at the same time (hard).
+* A student can attend at most one lesson at the same time (hard).
+* A teacher prefers to teach in a single room (soft).
+* A teacher prefers to teach sequential lessons and dislikes gaps between lessons (soft).
+* A student dislikes sequential lessons on the same subject (soft).
+
+== Quickstart structure
+
+image::projectStructure.png[Project structure]
+
+The project consists of three modules:
+
+* *amq-quarkus-school-timetabling-common* defines the problem domain, and the `SolverRequest` and `SolverResponse` classes for messaging.
+The following two modules depend on this one.
+* *amq-quarkus-school-timetabling-client* is the Client Quarkus application that contains a UI, a REST endpoint and a demo data generator.
+* *amq-quarkus-school-timetabling-solver* is the Solver Server Quarkus application that solves school timetabling problem instances coming via a message queue `solver_request`.
+
+== Messaging
+
+image::architecture.png[Architecture]
+
+The Client application serializes an unsolved `TimeTable` wrapped by the `SolverRequest` class into a JSON and sends it to the `solver_request` queue.
+The Solver Server receives the request from this queue, deserializes it and solves the `TimeTable` via OptaPlanner.
+After the solving finishes, the Solver Server wraps the `TimeTable` by the `SolverResponse` class, serializes it to a JSON and sends it to the `solver_response` queue.
+
+=== Requirements
+
+. No solver request message must be lost, even if the Solver Server crashes.
+. Any error that occurs in the Solver Server must be propagated back to the Client.
+. Invalid solver request message is sent to a dead letter queue.
+
+=== AMQ vs. Kafka
+
+One of the technologies being talked about these days is Kafka.
+It’s been used in https://medium.com/@rinu.gour123/apache-kafka-use-cases-and-applications-7a5f379b8451[various use cases], including message queues, replacing the traditional messaging systems.
+However, as much as Kafka might shine in its most typical use, delivering huge amounts of small messages, this is not the problem we are trying to solve in this post.
+There will not be millions of input problems to solve every second, but odds are problems will not be small.
```

Comment:
I'd avoid the usage of future tense even here and go with present tense.
The "will" is used here out of humbleness or not sure of scientific correctness.
But you define the use case. In this use case, there are no millions of input problems and there the problems are often big.